h2o-core/src/main/java/hex/ModelBuilder.java

```diff
@@ -308,21 +317,28 @@ public void run() {
       _mb.trainModel();
     }
   }
-
-  /** Method to launch training of a Model, based on its parameters. */
   final public Job<M> trainModel() {
+    return trainModel(null);
+  }
+  /** Method to launch training of a Model, based on its parameters. */
+  final public Job<M> trainModel(final Consumer<ModelBuildingResult> callback) {
     if (error_count() > 0)
       throw H2OModelBuilderIllegalArgumentException.makeFromBuilder(this);
     startClock();
     if( !nFoldCV() )
-      return _job.start(trainModelImpl(), _parms.progressUnits(), _parms._max_runtime_secs);
+      return _job.start(trainModelImpl(callback), _parms.progressUnits(), _parms._max_runtime_secs);
 
     // cross-validation needs to be forked off to allow continuous (non-blocking) progress bar
     return _job.start(new H2O.H2OCountedCompleter() {
                         @Override
                         public void compute2() {
                           computeCrossValidation();
                           tryComplete();
+                          if(callback != null) {
+                            final ModelBuildingResult modelBuildingResult = new ModelBuildingResult(Optional.of(_job.get()),
```

Comment:
this API seems strange because it seems like there are 2 use cases handled by a single method:

1. successful completion
2. exceptional completion

ModelBuildingResult - first 2 arguments to this class seem mutually exclusive

why does the callback not have 2 methods for these 2 use cases?

Reply:
To only have one universal callback. We might have two as well - no problem.

Reply:
Introduced dual callback - one on failure and one on success.